### PR TITLE
[MPLUGIN-419] Allow `@Parameter` on setters methods

### PR DESCRIPTION
**Pull request by slawekjaranowski**
- new implementation
- remove some of unused or unneeded methods

### CODE REVIEW

**Review by kwin — CHANGES_REQUESTED** [review]

**Comment by kwin** [comment]
Would be good to update https://maven.apache.org/guides/plugin/guide-java-plugin-development.html#parameters subsequently.

**Comment by slawekjaranowski** [comment]
> Would be good to update https://maven.apache.org/guides/plugin/guide-java-plugin-development.html#parameters and https://maven.apache.org/developers/mojo-api-specification.html subsequently.

Agree - documentation should be updated .... we need to review it after release

File: maven-plugin-plugin/src/site/apt/examples/using-annotations.apt.vm [hunk 1]
```diff
@@ -41,18 +41,22 @@ Using Plugin Tools Java Annotations
...
+  * 2 class level annotations:
 
     * <<<@Mojo>>>: This annotation will mark your class as a Mojo,
 
     * <<<@Execute>>>: Used if your Mojo needs to fork a lifecycle,
 
     []
 
-  * 2 field-level annotations:
+  * 1 filed or method level annotations:
```
[the remark this answers is not in the document]
Comment (kwin) [reply maven-plugin-plugin/src/site/apt/examples/using-annotations.apt.vm:52]: How was this and my other remarks solved? I don't see a new commit...

[the remark this answers is not in the document]
Comment (slawekjaranowski) [reply maven-plugin-plugin/src/site/apt/examples/using-annotations.apt.vm]: ups ... I forgot to push ... now should be visible 😄

**Review by kwin — APPROVED** [review]

**Comment by slawekjaranowski** [comment]
@kwin thanks for approve.
Please look again - I rebased code with your last change.

**Comment by kwin** [comment]
> @kwin thanks for approve. Please look again - I rebased code with your last change.

Still looks good ;-)

File: maven-plugin-tools-annotations/src/main/java/org/apache/maven/tools/plugin/extractor/annotations/scanner/MojoAnnotationsScanner.java [hunk 2]
```diff
@@ -45,6 +45,9 @@
                                                           Component.class.getName(),
                                                           Deprecated.class.getName() );
 
+    List<String> METHOD_LEVEL_ANNOTATIONS = Arrays.asList( Parameter.class.getName(),
+                                                          Deprecated.class.getName() );
```
Comment (gnodet) [inline maven-plugin-tools-annotations/src/main/java/org/apache/maven/tools/plugin/extractor/annotations/scanner/MojoAnnotationsScanner.java:49]: Column alignment for `Deprecated` ?

**Comment by slawekjaranowski** [comment]
> Looks good (apart from the very minor indentation issue). Some of it had already been integrated in the v4 api through [gnodet/maven@6317b8c](https://github.com/gnodet/maven/commit/6317b8c0ad468a4abdee47531d461755ff051d0a), but both `@Parameter` and `@Component` annotations were actually modified. Does it make sense or should I revert the changes on `@Component` ?

We can add support for `@Compoent` with method in next step.
I must check if old `@Compoent` can be injected by method - maybe you know it.

I hope we can also add support for old and new annotation together. 
As soon Maven 4 alpha will be published we will have open way for it.

**Comment by gnodet** [comment]
> > Looks good (apart from the very minor indentation issue). Some of it had already been integrated in the v4 api through [gnodet/maven@6317b8c](https://github.com/gnodet/maven/commit/6317b8c0ad468a4abdee47531d461755ff051d0a), but both `@Parameter` and `@Component` annotations were actually modified. Does it make sense or should I revert the changes on `@Component` ?
> 
> We can add support for `@Compoent` with method in next step. I must check if old `@Compoent` can be injected by method - maybe you know it.

I don't think so at the moment.  The question is more whether sisu-inject supports injecting components using setters, but I would think so.

> I hope we can also add support for old and new annotation together. As soon Maven 4 alpha will be published we will have open way for it.

That's done in https://github.com/apache/maven-plugin-tools/pull/117 . Both annotation packages are supported.    It's ready for review (may need to rebase based on recent changes), but I think it would be better in a new major version of the plugin tools.

**Pull request merged**